File: MPI_Code/Fortran/scatter.f90
PROGRAM scatter
USE mpi_f08
IMPLICIT NONE
INTEGER::i, isend(3),nprocs,myrank,irecv
CALL MPI_Init
CALL MPI_Comm_rank(MPI_COMM_WORLD,myrank)
CALL MPI_Comm_size(MPI_COMM_WORLD,nprocs)
IF(myrank==0)THEN
  DO i=1,nprocs
    isend(i)=i
  END DO
END IF
CALL MPI_Scatter(isend,1,MPI_INTEGER,irecv,1,MPI_INTEGER,0,MPI_COMM_WORLD)
PRINT*,'rank:',myrank,'   irecv=',irecv

isend=0
IF(myrank==0)THEN
  DO i=1,nprocs
    isend(i)=i*10
  END DO
END IF
irecv=0
IF(myrank==0)THEN
  irecv=isend(1)
  CALL MPI_Scatter(isend,1,MPI_INTEGER,MPI_IN_PLACE,1,MPI_INTEGER,0,MPI_COMM_WORLD)
ELSE
  CALL MPI_Scatter(isend,1,MPI_INTEGER,irecv,1,MPI_INTEGER,0,MPI_COMM_WORLD)
END IF
PRINT*,'RANK:',myrank,'   IRECV=',irecv

CALL MPI_Finalize
END PROGRAM scatter

File: MPI_Code/C/scatter.c
#include <stdio.h>
#include <mpi.h>
int main()
{
  int i, myrank;
  int isend[3]={0,}, irecv=0;
  MPI_Init(NULL,NULL);
  MPI_Comm_rank(MPI_COMM_WORLD,&myrank);
  if(myrank==0) for(i=0;i<3;i++) isend[i]=i+1;
  MPI_Scatter(isend,1,MPI_INT,&irecv,1,MPI_INT,0,MPI_COMM_WORLD);
  printf("myrank: %d  irecv=%d\n",myrank,irecv);

  if(myrank==0){
    for(i=0;i<3;i++) isend[i]=(i+1)*10;
  }
  else{
    for(i=0;i<3;i++) isend[i]=0;
  }
  irecv=0;
  if(myrank==0){
    irecv=isend[0];
    MPI_Scatter(isend,1,MPI_INT, MPI_IN_PLACE, 1, MPI_INT, 0, MPI_COMM_WORLD);
  }
  else{
    MPI_Scatter(isend,1,MPI_INT, &irecv,1, MPI_INT, 0, MPI_COMM_WORLD);
  }
  printf("RANK: %d,  IRECV=%d\n",myrank,irecv);
  MPI_Finalize();
  return 0;
}
